### PR TITLE
CxxValgrindSensor: allow multiple <stack> nodes #1440

### PR DESCRIPTION
**Pull request by ivangalkin**
* proper parsing of multiple <stack>s and <auxwhat>s
see https://github.com/pathscale/valgrind-mmt/blob/master/docs/internals/xml-output.txt

Each stackstrace of Valgrind will be stored as a separate NewIssue.
The reason is missing formatting for error messages
(even line breaks are not allowed).
Error messages with one stack trace are already hard to read.

* revisit Valgrind*::equals() and Valgrind*::hashCode() implementation
a. equals() shouldn not be implemented through hashCode() because
   there could be hash collisions and unequal objects could return
   true
b. perform equals()/hashCode() over all fields
   I believe that SonarQube has to show the Valgrind's report as-is.
   There should be no filtering or obfuscation

---
This change is [Reviewable](https://reviewable.io/reviews/sonaropencommunity/sonar-cxx/1461)

### CODE REVIEW

**Comment by ivangalkin** [comment]
This change was supposed to solve #1440 and be merged before #1436 
However there are some repeating CI problems

```
C:\Python27\Scripts\pip.exe install requests
Collecting requests
Exception:
Traceback (most recent call last):
  File "c:\python27\lib\site-packages\pip\basecommand.py", line 215, in main
    status = self.run(options, args)
  File "c:\python27\lib\site-packages\pip\commands\install.py", line 324, in run
    requirement_set.prepare_files(finder)
  File "c:\python27\lib\site-packages\pip\req\req_set.py", line 380, in prepare_files
    ignore_dependencies=self.ignore_dependencies))
  File "c:\python27\lib\site-packages\pip\req\req_set.py", line 620, in _prepare_file
    session=self.session, hashes=hashes)
  File "c:\python27\lib\site-packages\pip\download.py", line 821, in unpack_url
    hashes=hashes
  File "c:\python27\lib\site-packages\pip\download.py", line 659, in unpack_http_url
    hashes)
  File "c:\python27\lib\site-packages\pip\download.py", line 853, in _download_http_url
    stream=True,
  File "c:\python27\lib\site-packages\pip\_vendor\requests\sessions.py", line 488, in get
    return self.request('GET', url, **kwargs)
  File "c:\python27\lib\site-packages\pip\download.py", line 386, in request
    return super(PipSession, self).request(method, url, *args, **kwargs)
  File "c:\python27\lib\site-packages\pip\_vendor\requests\sessions.py", line 475, in request
    resp = self.send(prep, **send_kwargs)
  File "c:\python27\lib\site-packages\pip\_vendor\requests\sessions.py", line 596, in send
    r = adapter.send(request, **kwargs)
  File "c:\python27\lib\site-packages\pip\_vendor\cachecontrol\adapter.py", line 47, in send
    resp = super(CacheControlAdapter, self).send(request, **kw)
  File "c:\python27\lib\site-packages\pip\_vendor\requests\adapters.py", line 482, in send
    raise RetryError(e, request=request)
RetryError: HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Max retries exceeded with url: /packages/49/df/50aa1999ab9bde74656c2919d9c0c085fd2b3775fd3eca826012bef76d8c/requests-2.18.4-py2.py3-none-any.whl (Caused by ResponseError('too many 503 error responses',))
You are using pip version 9.0.1, however version 10.0.0 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.
Command exited with code 2
```

**Comment by ivangalkin** [comment]
Hi All! The change finally passes the CI build (I've needed [3] & [4] in order to retrigger it). Feel free to review. Thank you!

**Comment by guwirth** [comment]
@ivangalkin thx for providing this.

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindError.java [hunk 1]
```diff
@@ -29,54 +32,78 @@
...
+        .isEquals();
   }
 
   @Override
   public int hashCode() {
-    return new HashCodeBuilder()
-      .append(kind)
-      .append(stack)
-      .toHashCode();
+    return new HashCodeBuilder() //
```
Comment (guwirth) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindError.java]: empty comments at the end?

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindFrame.java [hunk 2]
```diff
@@ -77,25 +67,33 @@ public String toString() {
...
-    if (o == null || getClass() != o.getClass()) {
+    if (o == null)
+      return false;
+    if (getClass() != o.getClass())
       return false;
-    }
     ValgrindFrame other = (ValgrindFrame) o;
-    return hashCode() == other.hashCode();
+    return new EqualsBuilder() //
+        .append(ip, other.ip) //
```
Comment (guwirth) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindFrame.java]: empty comments at the end?

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindFrame.java [hunk 3]
```diff
@@ -77,25 +67,33 @@ public String toString() {
...
   public int hashCode() {
-    return new HashCodeBuilder()
-      .append(obj)
-      .append(fn)
-      .append(dir)
-      .append(file)
-      .append(line)
-      .toHashCode();
+    return new HashCodeBuilder()//
+        .append(ip) //
```
Comment (guwirth) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindFrame.java]: empty comments at the end?

**Comment by guwirth** [comment]
@ivangalkin if possible please also rebase it.

**Comment by guwirth** [comment]
@ivangalkin thx

**Comment by ivangalkin** [comment]
@guwirth I can rework #1447 now; I will also apply the ideas mentioned in #1440

**Pull request merged**